src/libraries/System.Linq/src/System/Linq/AnyAll.cs

```diff
@@ -57,6 +57,16 @@ public static bool Any<TSource>(this IEnumerable<TSource> source, Func<TSource,
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.predicate);
             }
 
+            if (source is List<TSource> list)
+            {
+                return !list.TrueForAll(element => !predicate(element));
+            }
+
+            if (source is TSource[] array)
+            {
+                return !Array.TrueForAll(array, element => !predicate(element));
+            }
```

Comment:
Thanks for the contribution, but this has a few issues.
1) It's capturing `predicate`, which is going to result in additional closure and delegate allocations.  And the way it's written, those allocations will occur even if the source isn't a list or an array.
2) This adds two type checks of overhead in the case where the source isn't a list or an array.  This might help in the case where it's a list or an array, but it will hurt in the case where it's not.  We need to weigh such trade-offs very carefully, especially considering small inputs.

Also, in general, perf-focused PRs should include benchmarks highlighting the impact of the change (positive and negative).

Thanks.